presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduSplitSource.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.kudu;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.spi.connector.ConnectorPartitionHandle;
+import io.prestosql.spi.connector.ConnectorSplitSource;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static java.util.function.Function.identity;
+
+public class KuduSplitSource
+        implements ConnectorSplitSource
+{
+    private final Map<Integer, KuduSplit> splitMap;
+    private AtomicInteger fetchedBucketCount = new AtomicInteger();
+
+    KuduSplitSource(List<KuduSplit> splits)
+    {
+        this.splitMap = requireNonNull(splits, "splits are null").stream()
+                .collect(toImmutableMap(KuduSplit::getBucketNumber, identity()));
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHandle partitionHandle, int maxSize)
+    {
+        KuduPartitionHandle kuduPartitionHandle = (KuduPartitionHandle) partitionHandle;
+        KuduSplit kuduSplit = splitMap.get(kuduPartitionHandle.getBucket());
+        fetchedBucketCount.incrementAndGet();
```

Comment:
the best would be to remove entries from `splitMap`. Notice that this code is vulnerable, in case you are asked for the bucket you might return same split twice and say in `isFinished` that you finished even if not all splits were procesed.

Reply:
Ok, I will do the change!

Reply:
Ok, I will do the change!